tests/unit/miniapps/test_tmop.cpp

```diff
@@ -0,0 +1,862 @@
+
+      newton->Mult(b, x.GetTrueVector());
+      x.SetFromTrueVector();
+
+      REQUIRE(newton->GetConverged());
+
+      const double final_energy = nlf.GetParGridFunctionEnergy(x);
+      res.final_energy = final_energy;
+   }
+
+   Vector &x_t(x.GetTrueVector());
+   double x_t_dot = x_t*x_t, dot;
+   MPI_Allreduce(&x_t_dot, &dot, 1, MPI_DOUBLE, MPI_SUM, pmesh->GetComm());
+   res.dot = dot;
+
+   delete S;
+   delete S_prec;
+   delete pmesh;
+   delete metric;
+   delete coeff1;
+   delete metric2;
+   delete newton;
+   delete target_c;
+   delete target_c2;
+   delete adapt_coeff;
+
+   return 0;
+}
+
+} // namespace mfem
+
+static int argn(const char *argv[], int argc =0)
+{
+   while (argv[argc]) { argc+=1; }
+   return argc;
+}
+
+static void req_tmop(int id, const char *args[], Req &res)
+{ REQUIRE(tmop(id, res, argn(args), const_cast<char**>(args))==0); }
+
+#define DEFAULT_ARGS const char *args[] = { \
+    "tmop_tests", "-pa", "-m", "mesh", "-o", "0", "-rs", "0", \
+    "-mid", "0", "-tid", "0", "-qt", "1", "-qo", "0", \
+    "-ni", "10", "-rtol", "1e-8", "-ls", "2", "-li", "100", \
+    "-lc", "0", "-nor", "0", "-ji", "0", "-nl", "1", "-cmb", "0", nullptr }
+constexpr int ALV = 1;
+constexpr int MSH = 3;
+constexpr int POR = 5;
+constexpr int RS = 7;
+constexpr int MID = 9;
+constexpr int TID = 11;
+constexpr int QTY = 13;
+constexpr int QOR = 15;
+constexpr int NI  = 17;
+constexpr int LS  = 21;
+constexpr int LI  = 23;
+constexpr int LC  = 25;
+constexpr int NOR = 27;
+constexpr int JI  = 29;
+constexpr int NL  = 31;
+constexpr int CMB = 33;
+
+static void dump_args(int id, const char *args[])
+{
+   if (id != 0) { return; }
+   const char *format =
+      "tmop -m %s -o %s -qo %s -mid %s -tid %s -ls %s"
+      "%s%s%s%s"         // Optional args: RS, QTY
+      "%s%s%s%s%s%s%s%s" // Optional args: LC, NOR, JI, NL & CMB
+      " %s\n";           // Assembly level
+   printf(format,
+          args[MSH], args[POR], args[QOR], args[MID], args[TID], args[LS],
+          // Optional args: RS, QTY
+          args[RS][0] == '0' ? "" : " -rs ",
+          args[RS][0] == '0' ? "" : args[RS],
+          args[QTY][0] == '1' ? "" : " -qt ",
+          args[QTY][0] == '1' ? "" : args[QTY],
+          // Optional args: LC, NOR, JI, NL & CMB
+          args[LC][0] == '0' ? "" : " -lc ",
+          args[LC][0] == '0' ? "" : args[LC],
+          args[NOR][0] == '0' ? "" : " -nor",
+          atof(args[JI]) == 0.0 ? "" : " -jitter",
+          args[NL][0] == '1' ? "" : " -nl ",
+          args[NL][0] == '1' ? "" : args[NL],
+          args[CMB][0] == '0' ? "" : " -cmb ",
+          args[CMB][0] == '0' ? "" : args[CMB],
+          // Assembly level
+          args[ALV]);
+   fflush(0);
+}
+
+static void tmop_require(int id, const char *args[])
+{
+   Req res[2];
+   (args[ALV] = "-pa", dump_args(id, args), req_tmop(id, args, res[0]));
+   (args[ALV] = "-no-pa", dump_args(id, args), req_tmop(id, args, res[1]));
```

Comment:
Since the goal is to compare results with `-pa` and `-no-pa`, I think better test names are`test_tmop_pa.cpp` and `[TMOP_PA]`.

Reply:
`done`